src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,176 @@
     }
 
-    @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Invalid index: " + index);
+        }
+
+        if (index == size) {
+            Node<T> oldLast = last;
+            Node<T> newNode = new Node<>(oldLast, value, null);
+            last = newNode;
+            if (oldLast == null) {
+                first = newNode;
+            } else {
+                oldLast.next = newNode;
+            }
+        } else if (index == 0) {
+            Node<T> oldFirst = first;
+            Node<T> newNode = new Node<>(null, value, oldFirst);
+            first = newNode;
+            if (oldFirst == null) {
+                last = newNode;
+            } else {
+                oldFirst.prev = newNode;
+            }
+        } else {
+            Node<T> currentNode = findNodeByIndex(index);
+            Node<T> prevNode = currentNode.prev;
+            Node<T> newNode = new Node<>(prevNode, value, currentNode);
+            prevNode.next = newNode;
+            currentNode.prev = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        if (list == null) {
+            throw new RuntimeException("Can't add list because it is null");
+        }
+        if (list.size() == 0) {
+            return;
+        }
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        indexValidation(index);
+        Node<T> currentNode = findNodeByIndex(index);
+        return currentNode.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        if (value == null) {
+            throw new IllegalArgumentException("Cannot set the value because it is null");
+        }
+        indexValidation(index);
+        Node<T> currentNode = findNodeByIndex(index);
+        T oldValue = currentNode.value;
+        currentNode.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        indexValidation(index);
+        Node<T> currentNode = findNodeByIndex(index);
+        T oldValue = currentNode.value;
+        unlink(currentNode);
+        return oldValue;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> currentNode = first;
+        if (object == null) {
+            while (currentNode != null) {
+                if (currentNode.value == null) {
+                    unlink(currentNode);
+                    return true;
+                }
+                currentNode = currentNode.next;
+            }
+        } else {
+            while (currentNode != null) {
+                if (object.equals(currentNode.value)) {
+                    unlink(currentNode);
+                    return true;
+                }
+                currentNode = currentNode.next;
+            }
+        }
```

Comment:
remove if-else
combine conditions 104,112 into a single null safe comparison condition, like in objects equals